Glslc supports setting shader resource limits

dneto0 opened the pull request.

dneto0 commented:
This PR will fix https://github.com/google/shaderc/issues/250

dneto0 commented:
Fixed a GCC warning about use without initialization. (Was a false positive, but GCC couldn't tell.)

antiagainst commented on glslc/README.asciidoc:
```diff
@@ -110,13 +115,41 @@ Otherwise,
+[[option-f-limit-file]]
+==== `-flimit-file`
+
+`-flimit-file <resource-limits-file>` lets you specify resource limits in a file.-
```
`-` at the end of the line. ;)

antiagainst commented on glslc/README.asciidoc:
```diff
@@ -110,13 +115,41 @@ Otherwise,
+
+`--show-limits` shows default resource limits for shader compilation.  The syntax
+is the same as accepted by `-flimit=` and for the contents of the file specified
+by `-flimit-file`.   This option accepts Glslang resource configuration files,
```
accepts? This option doesn't accept any additional parameters.

antiagainst commented on glslc/src/main.cc:
```diff
@@ -156,6 +197,34 @@ int main(int argc, char** argv) {
+      }
+      std::vector<char> contents;
+      if (!shaderc_util::ReadFile(limits_file.str(), &contents)) {
+        std::cerr << "glslc: Can't read limits file: " << limits_file
```
glslc: error:: cannot...

antiagainst commented on glslc/src/main.cc:
```diff
@@ -156,6 +197,34 @@ int main(int argc, char** argv) {
+      std::string err;
+      if (!SetResourceLimits(arg.substr(std::strlen("-flimit=")).str(),
+                             &compiler.options(), &err)) {
+        std::cerr << "glslc: -flimit error: " << err << std::endl;
```
glslc: error: ...

antiagainst commented on glslc/src/main.cc:
```diff
@@ -156,6 +197,34 @@ int main(int argc, char** argv) {
+              string_piece(contents.data(), contents.data() + contents.size())
+                  .str(),
+              &compiler.options(), &err)) {
+        std::cerr << "glslc: -flimit-file error: " << err << std::endl;
```
glslc: error: ...

antiagainst commented on glslc/src/resource_parse.cc:
```diff
@@ -0,0 +1,97 @@
+    bool ignore = IgnoreSetting(limit_name);
+    if (!ignore) {
+      if (!StringToLimit(limit_name, &limit))
+        return failure(std::string("Invalid resource limit: " + limit_name));
```
I'd prefer to keep all error messages with lower case letters b/c we don't know how it will be concatenated in the caller. It's strange to read an error message having multiple capitalized words in the middle.

But it's too trivial a thing.

antiagainst commented on glslc/test/option_flimit.py:
```diff
@@ -0,0 +1,126 @@
+                                            expect.CorrectObjectFilePreamble):
+    """Tests -flimit-file with lower than default argument.  The shader uses that offset."""
+
+    limits_file = FileShader('MinProgramTexelOffset -9', '.txt')
```
Well, this is a hack. ;)
Should use `File` in `environment` instead. See `option_dash_M.py` or something else for an example.

antiagainst commented on glslc/test/option_flimit.py:
```diff
@@ -0,0 +1,126 @@
+
+    limits_file = FileShader('MinProgramTexelOffset -9', '.txt')
+    shader = shader_with_tex_offset(-9);
+    glslc_args = [ '-c', shader, '-flimit-file', limits_file ]
```
redundant spaces around `[]`.

antiagainst commented on glslc/README.asciidoc:
```diff
@@ -110,13 +115,41 @@ Otherwise,
+`-flimit=<resource-limits>` lets you specify resource limits for the shader.
+The argument should be a sequence of limit name, integer value pairs.  Tokens
+should be separated by whitespace.  Use `--show-limits` to show the default values,
+and example syntax.
```
Does this affect shaders specified before it? Or does it behave similarly to `-fshader-stage`?

antiagainst commented on glslc/src/main.cc:
```diff
@@ -46,13 +49,21 @@ An input file of - represents standard input.
+  -flimit=<settings>
+                    Specify resource limits. Each limit is specified by a limit
+                    name followed by an integer value.  Tokens should be
+                    separated by whitespace.  If the same limit is specified
```
The last sentence should also appear in the manual, which is expected to provide more info than the help message.

dneto0 commented:
Rebased.  Added https://github.com/google/shaderc/pull/277/commits/e062cd4c276f4324b56324265865d2b84c1e13f5  and https://github.com/google/shaderc/pull/277/commits/33b51dc24c17549e7e64493780179dd5a12ec109 to address review feedback.

antiagainst commented on glslc/test/option_flimit.py:
```diff
@@ -0,0 +1,137 @@
+    limits_file = File('limits.txt', 'MinProgramTexelOffset -9')
+    shader = File('shader.vert', shader_source_with_tex_offset(-9));
+    environment = Directory('.', [limits_file, shader])
+    print("sHHHH is: " + shader.name)
```
Debuggie stuff. ;)

dneto0 replied:
Thank you.  :-)

dneto0 commented:
Removed the debug print, squash, and pushed into master as 2df47b51d83a
